### PR TITLE
fix(web): clean and markdown-render synthesize-mode memory recall card

### LINKED ISSUE #2142: Structured view: memory recall card shows raw <system-reminder> tags and line numbers, should render markdown
**Note:** Claude handled the investigation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Problem

When the structured view renders a synthesize-mode memory recall, the card body shows the recalled memory text raw and unprocessed. Three things are wrong with it:

1. It contains literal `<system-reminder>` / `</system-reminder>` wrapper tags from the SDK's recall payload.
2. It carries `cat -n` style line-number prefixes (e.g. `   12\tsome text`) on every line.
3. It is markdown, but the card renders it verbatim in a `<pre>`, so headings, lists, and links show as raw `#`, `-`, and `[...]()` source.

The memory recall card is supposed to be a clean, readable view of what the agent recalled. Right now it leaks the internal transport format into the UI.

## What's there today

The synthesize-mode body is rendered verbatim in a `<pre>` at `web/src/components/acp/ToolCards.tsx:1744-1750`:

```tsx
{isSynthesize ? (
  <pre
    data-testid="memory-recall-synthesized"
    className="whitespace-pre-wrap break-words text-[11px] text-text-secondary"
  >
    {synthesized}
  </pre>
) : (
```

`synthesized` is `recall.synthesized_text` (`ToolCards.tsx:1714`), passed through with no cleanup.

The text originates from the adapter. claude-agent-acp emits one text content block per recalled memory in synthesize mode, using each memory's raw `m.content` (`@agentclientprotocol/claude-agent-acp/dist/acp-agent.js:678-685`):

```js
const content = isSynthesis
  ? message.memories
      .filter((m) => typeof m.content === "string")
      .map((m) => ({ type: "content", content: { type: "text", text: m.content } }))
  : [];
```

aoe concatenates those blocks straight into `synthesized_text` with no sanitization (`src/acp/acp_client.rs:3714-3720`):

```rust
let synthesized_text = if mode == "synthesize" {
    let text = extract_tool_content_text(content);
    if text.is_empty() { None } else { Some(text) }
} else {
    None
};
```

So whatever the SDK put in `m.content`, including the `<system-reminder>` wrapper and `cat -n` line numbers, reaches the `<pre>` untouched.

The existing render test only exercises a clean fixture string (`web/src/components/acp/__fixtures__/toolCalls.ts:196` is `"User is a senior engineer working on agent-of-empires."`), so the dirty real-world shape was never covered.

## Root cause

Two layers, one fix point each:

- **Formatting (display):** `MemoryRecallCard` renders `synthesized_text` in a `<pre>` (`ToolCards.tsx:1744-1750`). It should render markdown via the existing `<Markdown>` component (`web/src/components/acp/Markdown.tsx`), the same one assistant text uses.
- **Sanitization (content):** nothing strips the `<system-reminder>` wrapper or the `cat -n` line-number prefixes before display. There is no cleanup at ingest (`acp_client.rs:3714`) nor at render (`ToolCards.tsx:1714`).

## Proposal

**(A) Minimum fix, web-only.** Sanitize and markdown-render in `MemoryRecallCard`:

1. Strip the `<system-reminder>` / `</system-reminder>` wrapper tags from `synthesized`.
2. Strip leading `cat -n` line-number prefixes (`^\s*\d+\t`) per line.
3. Replace the `<pre>{synthesized}</pre>` with `<Markdown text={cleaned} />`.

A small helper, e.g. `cleanRecalledMemory(text: string): string`, colocated in `memoryClassify.ts` or a new `memoryRecall.ts`, keeps the regex testable in isolation. This is the lazy, reversible choice: raw payload stays in the event store, the card just presents it clean.

**(B) Sanitize at ingest instead.** Do the stripping in `extract_memory_recall` (`acp_client.rs:3714`) so the persisted `synthesized_text` is already clean, then the card only needs the `<Markdown>` swap. Cleaner data model, but mutates persisted payloads and splits the logic across Rust + web. Prefer (A) unless we want the stored event sanitized too.

Recommend **(A)**. The wrapper/line-number shape is a display concern and the regex belongs next to the renderer.

## Why now

- Memory recall cards are a headline feature of the structured view (#1071, #1402); leaking transport format makes them look broken.
- The fix is small and self-contained (one component, one helper, one test).
- The `<Markdown>` component already exists, so there is no new dependency or infra.

## User stories / regression tests

Implementation MUST add tests covering each story below; PR is not done without them.

- As a user, when I expand a synthesize-mode memory recall card, then the body renders as formatted markdown (headings, lists, links), not raw `#`/`-`/`[]()` source.
  - Suggested test layer: Vitest + RTL on `MemoryRecallCard` (`web/src/components/acp/ToolCards.render.test.tsx`).
- As a user, when the recalled memory text is wrapped in `<system-reminder>...</system-reminder>`, then those literal tags do not appear in the rendered card.
  - Suggested test layer: Vitest unit on the `cleanRecalledMemory` helper.
- As a user, when the recalled memory text has `cat -n` line-number prefixes, then no leading line numbers appear in the rendered card.
  - Suggested test layer: Vitest unit on the `cleanRecalledMemory` helper.

## Open questions

- Sanitize at render (A) or at ingest (B)? Recommendation is (A); confirm before implementing.
- Should the same cleanup apply to the generic Read-of-memory `MemoryCard` path (#1071 path-sniff), or is this strictly the synthesize-mode `MemoryRecallCard`? The report is about memory recalls specifically; scoping to `MemoryRecallCard` first.

## Out of scope

- Native TUI memory-recall rendering. The TUI structured view has no dedicated memory-recall card today (`src/tui/structured_view/reducer.rs:578` only nulls the field); adding one is a separate gap.
- Changing what the adapter emits (upstream claude-agent-acp behavior).

## Cross-references

- #1071 — closed; introduced the dedicated memory recall/save card this bug lives in.
- #1402 — merged; added the `memory_recall` tool-channel handling and `synthesized_text` plumbing.

### PR DESCRIPTION
**Pull request by Seluj78**
**Note:** Claude handled the implementation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Description

The structured view's synthesize-mode memory recall card rendered `synthesized_text` verbatim in a `<pre>`. That leaked the SDK's transport format straight into the UI: the `<system-reminder>` envelope tags, `cat -n` line-number prefixes on every line, and markdown shown as raw `#`/`-`/`[]()` source.

This adds `cleanRecalledMemory` (in `web/src/lib/memoryClassify.ts`) to strip the `<system-reminder>` envelope and the per-line number prefixes, then renders the cleaned body as markdown via `marked`. The structured view's `<Markdown>` component needs the assistant-ui runtime provider and throws "requires an AuiProvider" outside the transcript (the same reason `CommentMarkdown` exists), so the card uses `marked` directly and reuses the existing `.acp-markdown` prose styling.

While adding an end-to-end test for this, I found a latent bug: the synthesize/recall `MemoryRecallCard` was unreachable in the live structured view. `AcpRuntime` flattens each tool call into an assistant-ui part that carries only toolName/args/result, dropping the structured `memory_recall` payload, so `AssistantToolCall` rebuilt a `ToolCall` without it and the dispatcher fell back to a generic read card. The existing test fed a preset fixture straight to `ToolCard`, so the gap was invisible. The second commit smuggles `memory_recall` through the part args (mirroring `_aoe_parent_tool_call_id`) and rebuilds it on the reconstructed tool, so the card now actually renders in the app.

User-observable impact: the memory recall card now shows clean, formatted markdown instead of wrapped, line-numbered raw text, and the dedicated card now renders at all in the live structured view.

Closes #2142

## PR Type

- [ ] New Feature
- [x] Bug Fix
- [ ] Refactor
- [ ] Documentation
- [ ] Infrastructure / CI

## How to test

- [ ] Open a structured-view session where a synthesize-mode memory recall card appears, expand the "Memory recall / Synthesised memory" card, and confirm no literal `<system-reminder>` text is shown.
- [ ] In that same card, confirm there are no leading `cat -n` line numbers on the body lines.
- [ ] Confirm the recalled memory renders as formatted markdown (headings, bullet lists) rather than raw `#` / `-` source.

## Checklist

- [x] New and existing tests pass
- [x] Documentation was updated where necessary
- [ ] For UI changes: included screenshot or recording

## Test Coverage Analysis

**Web dashboard / structured view (Playwright user story)**
- [x] Added a live Playwright spec `web/tests/live/acp-memory-recall.spec.ts` (mapped via the new `acp.memory-recall` entry in `web/tests/coverage-matrix.json`) that scripts a synthesize-mode `memory_recall` through the fake ACP agent and asserts the card renders cleaned markdown end-to-end. Also updated the `ToolCards.render.test.tsx` synthesize case and added unit tests for `cleanRecalledMemory` in `memoryClassify.test.ts`.

**TUI / CLI (e2e test)**
- [x] N/A: this PR doesn't change TUI rendering, a CLI subcommand, or session lifecycle

## AI Usage

- [ ] No AI was used
- [x] AI was used

**AI Model/Tool used:** Claude (via Claude Code, agent-of-empires `/aoe-implement` skill)

**Any Additional AI Details you'd like to share:**
Investigation, plan, and implementation drafted by Claude under my direction. I made the design calls (sanitize at render, `marked` over the assistant-ui `<Markdown>`), reviewed each change, and confirmed the reproduce-then-fix test goes red on the pre-fix tree.

**NOTE:**
When responding to reviewer questions, please respond yourself rather than copy/pasting reviewer comments into an AI and pasting back its answer. We want to discuss with you, not your AI :)

- [x] I am an AI Agent filling out this form (check box if true)

## Overview

This PR fixes the structured view’s “memory recall” cards—especially **synthesize mode**—so synthesized memories show as **clean, safe, formatted markdown** instead of leaking internal SDK transport text.

### Correct synthesize-mode rendering
- The synthesize-mode memory recall card no longer displays the raw `synthesized_text` verbatim in a `<pre>`.
- It now renders **formatted markdown** (so headings, lists, and links appear properly).

### Display-time cleanup of leaked formatting
- Added a `cleanRecalledMemory` helper to remove:
  - the literal `<system-reminder>` / `</system-reminder>` wrapper tags
  - `cat -n`-style line-number prefixes
- Cleaning happens at **render time** (no change to persisted payloads).

### Structured view now preserves and routes `memory_recall`
- Updated the structured view runtime/tool-call pipeline so the `memory_recall` payload isn’t dropped and the UI doesn’t fall back to a generic read card.
- Added parsing/extraction utilities for an `_aoe_memory_recall` “smuggled” payload so the renderer can reliably recover the structured `memory_recall` data.

### Safer markdown output
- Markdown is rendered via `marked`, and the resulting HTML is sanitized (using `DOMPurify`) before being inserted into the page.

## Tests Added/Updated

- **Unit tests** for `cleanRecalledMemory` to confirm wrapper stripping, line-number removal, and correct edge-case handling.
- **Component test updates** for ToolCards to verify:
  - wrapper tags and line-number prefixes are not rendered
  - markdown renders to real DOM elements (not raw source)
  - dangerous HTML is sanitized/blocked
- **End-to-end Playwright coverage** for the full structured-view pipeline, including a malformed-payload case to ensure correct fallback behavior.

## Benefits

- **Users see readable content** (proper markdown) instead of internal transport noise.
- **No leaked system markers** (`<system-reminder>`, line-number prefixes).
- **Right-card routing works** so synthesize-mode behaves consistently in structured view.
- **Improved safety** by sanitizing rendered markdown output.

### CODE REVIEW

File: web/src/components/acp/ToolCards.tsx [hunk 1]
```diff
@@ -1711,8 +1712,14 @@ function MemoryRecallCard({ tool, result }: Props) {
...
   const paths = recall.paths ?? [];
-  const synthesized = recall.synthesized_text ?? "";
+  const synthesized = cleanRecalledMemory(recall.synthesized_text ?? "");
   const isSynthesize = recall.mode === "synthesize";
+  // Render the cleaned memory body as markdown. We use `marked` rather
+  // than the structured-view `<Markdown>` component because that one
+  // requires the assistant-ui runtime provider (it throws "requires an
+  // AuiProvider" outside the transcript), the same reason CommentMarkdown
+  // exists. The `.acp-markdown` class reuses the panel's prose styling.
+  const synthesizedHtml = isSynthesize ? (marked.parse(synthesized, { async: false }) as string) : "";
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply web/src/components/acp/ToolCards.tsx:1727]: Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2165/commits/f1cf331c — the parsed `marked` output now runs through DOMPurify before `dangerouslySetInnerHTML`, and a regression test asserts `onerror` handlers and `javascript:` URLs are stripped from agent-supplied memory content.

File: web/src/components/acp/StructuredView.tsx [hunk 2]
```diff
@@ -636,6 +637,7 @@ function AssistantToolCall(props: ToolCallProps) {
     kind: props.toolName,
     args_preview: props.argsText ?? safeStringify(props.args ?? null),
     started_at: startedAt,
+    memory_recall: pickMemoryRecall(props.args, props.argsText),
   };
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply web/src/components/acp/StructuredView.tsx:649]: Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2165/commits/10de630f — `memory_recall` is now reconstructed in the grouped (`groupChildToItem`) and subagent child paths too, not just the top-level call.

File: web/src/components/acp/StructuredView.tsx [hunk 3]
```diff
@@ -680,6 +682,28 @@ function pickStartedAt(args: Record<string, unknown> | undefined, argsText: stri
...
+    try {
+      const parsed = JSON.parse(argsText) as Record<string, unknown>;
+      const mr = parsed?._aoe_memory_recall;
+      if (mr && typeof mr === "object") return mr as MemoryRecall;
+    } catch {
+      // ignore
+    }
+  }
+  return undefined;
+}
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply web/src/components/acp/StructuredView.tsx]: Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2165/commits/10de630f — `pickMemoryRecall` now validates and normalizes via `asMemoryRecall` (mode must be a string; non-string `paths`/`synthesized_text` are dropped), returning undefined for malformed payloads. Extracted to `web/src/lib/memoryRecall.ts` with exhaustive unit tests in a follow-up commit.

**Review by njbrake — APPROVED** [review]
